Multiple promotion chats and reduce pylint warns

dashezup opened the pull request:
- Option for deleting join messages
- Check if user is member of any promotion chats (group/channel) instead of just one

I've pushed commit for this two, after was settled, I'll start try to reduce pylint warns (mostly no-self-use and no-else-return)

[Pylint features](http://pylint.pycqa.org/en/latest/technical_reference/features.html)

dashezup commented:
Should we also limit the usage of the bot as captcha bot in specified groups? @konnovdev seems it will make `config.py` being more complicated, tuples includes tuples and includes another tuples.

konnovdev commented:
`Should we also limit the usage of the bot as captcha bot in specified groups? @konnovdev seems it will make config.py being more complicated, tuples includes tuples and includes another tuples.`

I haven't thought of that, but for now the captcha feature will only work if the bot is an admin. so you can say it's configurable, you can turn it on and off by promoting/demoting bot

konnovdev commented:
btw just noticed, in `text handler` method `__is_provide_service` has to be renamed to `__should_provide_service`, if we want to keep english grammar correct

dashezup commented:
> for now the captcha feature will only work if the bot is an admin (we check with method self.__is_bot_admin(message.chat.id)). so you can say it's configurable, you can turn it on and off by promoting/demoting bot

ok, seems we don't need to change.
fwiw, there could be case that a bot get HTTP code 429 from telegram when it sends too many requests when it's in a group where slow mode was on but it dose not check the waiting time. It won't be the case with this bot because it only works when it's admin and admin does not being limited with slow mode. and it's very unlikely to reach the rate limit.

> btw just noticed, in text handler method __is_provide_service has to be renamed to __should_provide_service, if we want to keep english grammar correct

yeah. that confused me for a while, I feel that was wrong but didn't figure out. pushed new commit and fixed this.

dashezup commented:
Packaged it and updated README

The `pylintrc` is generated by `pylint --generate-rcfile >pylintrc`.
added `e` (error) and `f` (file) in "good-names", disabled the following warns:

```
C0114, # missing-module-docstring
C0115, # missing-class-docstring
C0116, # missing-function-docstring
``` 

<details>
  <summary>pylint output</summary>
  
```
$ venv/bin/pylint cpop_bot
************* Module cpop_bot.command_handler
cpop_bot/command_handler.py:9:0: R0903: Too few public methods (1/2) (too-few-public-methods)
************* Module cpop_bot.music_grabber
cpop_bot/music_grabber.py:55:4: R0201: Method could be a function (no-self-use)
cpop_bot/music_grabber.py:61:4: R0201: Method could be a function (no-self-use)
cpop_bot/music_grabber.py:86:4: R0201: Method could be a function (no-self-use)
cpop_bot/music_grabber.py:89:4: R0201: Method could be a function (no-self-use)
cpop_bot/music_grabber.py:96:4: R0201: Method could be a function (no-self-use)
cpop_bot/music_grabber.py:112:4: R0201: Method could be a function (no-self-use)
cpop_bot/music_grabber.py:103:0: R0903: Too few public methods (1/2) (too-few-public-methods)
************* Module cpop_bot.new_members_handler
cpop_bot/new_members_handler.py:12:0: R0903: Too few public methods (0/2) (too-few-public-methods)
cpop_bot/new_members_handler.py:56:4: R0201: Method could be a function (no-self-use)
cpop_bot/new_members_handler.py:60:4: R0201: Method could be a function (no-self-use)
************* Module cpop_bot.text_handler
cpop_bot/text_handler.py:56:4: R0914: Too many local variables (21/15) (too-many-locals)
cpop_bot/text_handler.py:100:15: W0703: Catching too general exception Exception (broad-except)
cpop_bot/text_handler.py:109:11: R0916: Too many boolean expressions in if statement (6/5) (too-many-boolean-expressions)
cpop_bot/text_handler.py:108:4: R0201: Method could be a function (no-self-use)
cpop_bot/text_handler.py:116:4: R0201: Method could be a function (no-self-use)

------------------------------------------------------------------
Your code has been rated at 9.47/10 (previous run: 9.47/10, +0.00)
```
</details>


I only made it not warn `logging-not-lazy`, `wrong-import-order`, `no-else-return`, `no-else-raise` and `invalid-name`.
I don't know how to deal with "no-self-use" and "too-few-public-methods", i think I don't know how to properly use class and methods.

konnovdev commented on cpop_bot/command_handler.py:
```diff
@@ -16,7 +16,7 @@ async def handle(self, bot_username, message: types.Message):
             await self.__handle_start(message)
         else:
-            logger.debug("USER TYPED UNKNOWN COMMAND " + message.text)
+            logger.debug("USER TYPED UNKNOWN COMMAND %s", message.text)
```
is using %s a better practice in this, or why did you decide to change it

konnovdev replied:
it's interesting because this way of making an output
```
"USER TYPED UNKNOWN COMMAND %s", message.text
```
is actually used in **C**
Alright if pylint says it's better then i guess it should be right

konnovdev commented on cpop_bot/music_grabber.py:
```diff
@@ -39,27 +39,24 @@ def download(self, input_url, filesize_limit):
+            self.__delete_downloaded_files(info['downloads'])
+            raise ValueError(webpage_info +
+                             ": failed to download audio or thumbnail")
+        raise WrongFileSizeError(webpage_info + ": audio file size (" +
```
it looks weird when you do it this way as if it's a default behivour. If you don't want to have **else** i think it's better to reverse the order and raise these errors right after the checks failed, so what i mean is do something like this:
```
if int(audio_filesize) > filesize_limit: 
          raise WrongFileSizeError(....)
....
# do your normal code of processing the audio
```

and then 
```
 if !self.__files_successfully_downloaded(info['downloads']):
         self.__delete_downloaded_files(info['downloads'])
         raise ValueError(webpage_info +
                             ": failed to download audio or thumbnail")
 return info
```

dashezup replied:
force pushed the new commit which solved this.

dashezup replied:
oh btw `!_self...` does not works, need to use `if not self...`

konnovdev commented on cpop_bot/text_handler.py:
```diff
@@ -36,16 +36,17 @@ async def handle(self, message: types.Message):
+                    return True
                 await message.reply(AVAILABILITY_HTML,
                                     parse_mode=ParseMode.HTML)
                 return False
```
looks like you can remove some returns. 

Right now your code looks like this
```
    async def __should_provide_service(self, message: types.Message):
        chat_id = message.chat.id
        if chat_id in WHITELIST_CHAT_ID:
            return True
        chat_type = message.chat.type
        if chat_type == "private":
            user_id = chat_id
            for promotion_chat in PROMOTION_LIST_CHAT_ID:
                if (await self.bot.get_chat_member(
                        promotion_chat, user_id)).is_chat_member():
                    return True
                await message.reply(AVAILABILITY_HTML,
                                    parse_mode=ParseMode.HTML)
                return False
        else:
            return False
```

You can change it to this:
```
    async def __should_provide_service(self, message: types.Message):
        chat_id = message.chat.id
        if chat_id in WHITELIST_CHAT_ID:
            return True
        chat_type = message.chat.type
        if chat_type == "private":
            user_id = chat_id
            for promotion_chat in PROMOTION_LIST_CHAT_ID:
                if (await self.bot.get_chat_member(
                        promotion_chat, user_id)).is_chat_member():
                    return True
                await message.reply(AVAILABILITY_HTML,
                                    parse_mode=ParseMode.HTML)
        return False
```

dashezup replied:
That's a for loop inside if else statement, not just if inside if.

with your suggestion, if the user is not member of any chat in PROMOTION_LIST_CHAT_ID, I guess the the function will return None in the final loop.

with the current commit in the PR, the for loop returns "True" or "False", if it always return True or False, then we can remove the else on the first level. But I'm not sure if it always return True/False (let's assume that in some extreme case that the for loop fails before return?)

also pylint does not warn about that. there is no "no-else-return" warns for now.

dashezup replied:
right, the for loop shouldn't return False but return False at the end of the function. new commit solved this.

konnovdev commented on cpop_bot/text_handler.py:
```diff
@@ -36,16 +36,17 @@ async def handle(self, message: types.Message):
-            if member.is_chat_member():
-                return True
-            else:
+            user_id = chat_id
```
are you sure that **user_id** is **chat_id**?

dashezup replied:
Yes, when the "chat_type" is `private`, the message.from.id is the same value as message.chat.id which is the user's id (the one who was interacting with the bot in private).

it's would be more confusing if it's `(await self.bot.get_chat_member(promotion_chat, chat_id)`

I added the variable to avoid confusing with that.

dashezup replied:
we may could get message.from.id and assign it to the variable "user_id" but I didn't figure out. plus we already have "chat_id" and it's not very confusing that when it's private chat that the chat_id is basically just the user id.

konnovdev replied:
oh makes sense, i didn't know that chatId = message.from.id in private. Yeah it's fine to leave it like this then

konnovdev commented on text_handler.py:
```diff
@@ -42,13 +42,15 @@ async def __is_provide_service(self, message: types.Message):
+                else:
+                    await message.reply(AVAILABILITY_HTML,
+                                        parse_mode=ParseMode.HTML)
+                    return False
```
i mean why would we need to add the bot in the channel anyway?

dashezup replied:
This probably also applies to the group, if the PROMOTION_LIST_CHAT_ID has a group where the bot is not a member, it will also fails to get members list and output errors (won't terminate the bot, it will keep running fine but just won't send AVAILABILITY_HTML message)

In your case, you probably don't need to add a channel id to PROMOTION_LIST_CHAT, but only the group id one or more ones.

iirc telegram channels and groups are basically same for telegram servers, they just have different permissions. in channels that only admin can see list of members and members can't send messages. in group, every member could see member list and can send message by default.

in my case, when I need to share the bot with people, it's not necessarily to join the group, but join a channel instead (it's also good that they won't see each other in member list). it's also a good way to promote the channel.

dashezup commented:
seems fine now, tested and merge so that I can work on the next PR.
